Automattic/wp-calypso
# Plans: Introduce and hook up Jetpack product installer

This PR introduces a Jetpack product installer in the plans "My plan" page, where the new security checklist lives. This is aiming to replace the existing client-side Jetpack plugin setup process that happens after someone purchases a Jetpack plan. 

For more context: p5XAZ9-2c7-p2

This is a big PR - I'm happy to split out some of the functionality into multiple PRs, but for the sake of testing, and given the range of this task I wrapped it as a single PR.

#### Changes proposed in this Pull Request

This PR introduces state, data layer and a component for installing Jetpack products, along with some UI changes in the thank you card and some behavioral changes. Uses an existing endpoint and D25853-code for the installation on the server side. A detailed list of changes is as follows:

* Introduce action creators for:
  * Starting product install
  * Requesting and receiving current product install status
* Introduce reducer for product install status
* Introduce data layer for:
  * Starting product install
  * Requesting and receiving current product install status
* Introduce several new selectors:
  * `getPluginKeys` selector
  * `getJetpackProductInstallStatus` selector
  * `getJetpackProductInstallProgress` selector
* Update and cleanup tests premium plugin selector tests.
* Introduce Jetpack product installer component that:
  * Uses D25853-code and an existing endpoint for installing, activating and configuring our product plugins
  * Handles installation and configuration process.
  * Retries when it should and can, provides insights and next actions for the user when it can't proceed further.
* Update the thank you card in the My plan page with the new UI.
* Point the user after checkout to My plan page with the new plan setup and the checklist.
* Add tests for:
  * Product install status reducer
  * `getPluginKeys` selector
  * `getJetpackProductInstallStatus` selector
  * `getJetpackProductInstallProgress` selector

#### Testing instructions

* Apply D25853-code on your WP.com sandbox.
* Sandbox `public-api.wordpress.com`.
* Do either of the following:
  * Connect a fresh site with the latest Jetpack (add `&calypso_env=development` to the connection URL), and buy a plan, then observe when you reach the My Plan page.
  * Use an existing connected Jetpack site with a plan, but delete OR just deactivate Akismet and VaultPress before the next step, and go to http://calypso.localhost:3000/plans/my-plan/:siteSlug?thank-you where `:siteSlug` is the slug of your Jetpack site.
    * You can use WP CLI's `wp plugin deactivate akismet vaultpress --uninstall` to deactivate and uninstall the plugins, or `wp plugin deactivate akismet vaultpress` to just deactivate the plugins on the Jetpack site.
* Observe the installation progress.
* Verify that if setup fails, it retries several times before displaying an error message. You can edit the endpoint and force `vaultpress_status` to be `vaultpress_error` to witness this.
* Verify that if setup succeeds, you are presented with a success message.
* Verify that after a success message, Akismet and VaultPress have been setup for your site properly.

#### Preview

Starting:
![](https://cldup.com/fGUkJsZ_OU.png)

Progress:
![](https://cldup.com/lG2yplThtl.png)

More progress:
![](https://cldup.com/_w4RjPuxpi.png)

Even more progress:
![](https://cldup.com/THLJJy7wAi.png)

Success:
![](https://cldup.com/1CutDx70PL.png)

Persistent error and next actions:
![](https://cldup.com/2kA_56gb-E.png)
